fix: Fix MV2 build sourcemap upload

Gudahtt opened the pull request:
## **Description**

The sourcemaps for MV2 builds (used for Firefox) were not being uploaded to Sentry at all. This resulted in invalid stack traces for Firefox error reports.

The Sentry initiatization has been updated to use a `dist` option, letting us differentiate between different types of build for the same version. This is now used to signify which builds are mv2 and which are mv3. Both distributions are uploaded separately as part of the release process (for Flask and main builds; we don't have MV2 builds for MMI or beta).

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/26467?quickstart=1)

## **Related issues**

Fixes #26466

## **Manual testing steps**

Unfortunately I don't know of an easy way to test the CircleCI changes. Those will just have to be read carefully.

We can test the application changes and the Sentry script changes though. The two outcomes we want to test are:
* The source code and sourcemaps are uploaded correctly for both MV2 and MV3 builds
* Application bundles for MV2 and MV3 builds are reporting errors properly tagged as being from an MV2 or MV3 build, and are mapped correctly in Sentry.

Here are the steps I used to test this:
* Setup a personal Sentry account with a `metamask` project
* Create a Custom Integration in Sentry (Settings -> Custom Integrations)
  * Steps:
    * Navigate to "Settings -> Custom Integrations" on the Sentry dashboard for your personal Sentry account
    * Click the purple "Create New Integration" button on the top-right of the page
    * Select "Internal integration"
    * Provide any name, and grant it "Admin" access to "Releases"
  * Alternatively an Auth Token might work as well, but some of the commands used by our script to check for pre-existing releases seem to not work with the `org:ci` scope, which is the only scope available for Sentry auth tokens.
* Checkout this branch
* Bump the patch version in `package.json` (just in case you need to re-test, this is an easy way to separate old errors/builds from new ones)
* Before creating builds, make sure that `SENTRY_DSN_DEV` is set to the DSN of your personal Sentry account's `metamask` project
* Run `yarn dist:mv2` to create an MV2 build
* Move it to the `dist-mv2` directory (`mv dist dist-mv2`)
* Run `yarn dist` to create an MV3 build
* Before uploading sourcemaps, make sure that `SENTRY_AUTH_TOKEN` is set to the Auth Token generated from the custom integration earlier.
* Run `yarn sentry:publish --org [your organization]` to upload the MV3 build
* Run `yarn sentry:publish --org [your organization] --dist mv2` to upload the MV2 build

At this point, you should be able to see the releases on the Sentry dashboard along with the artifacts. Look in "Settings -> Projects -> Source Maps" for these. They are labeled by release number and dist.

Now, load each build in your browser (one at a time, never both enabled at once) and follow these steps:
* Proceed through onboarding, opting in to MetaMetrics
* Navigate to the test-dapp and connect to it
* Click the "INVALID TRANSACTION TYPE (NOT SUPPORTED)" button in the "Malformed Transactions" section of the test dapp, then reject the confirmation after it shows up. This should trigger an error in Sentry.
* Look for the error in Sentry and ensure that the frame that shows `transactionController.updateSecurityAlertResponse(` is mapped to source code correctly (it should resolve to `app/scripts/lib/ppom/ppom-util.ts`). Also check that the `dist` is correct in the "Tags" section of the Sentry issue page.

<details>
<summary> Here are screenshots of what that looked like for me: </summary>

![Screenshot 2024-08-20 at 19 19 15](https://github.com/user-attachments/assets/36ca0422-71f0-456a-8f1f-5ac980aa5fea)

![Screenshot 2024-08-20 at 19 06 08](https://github.com/user-attachments/assets/25fd681f-c933-41d5-be0e-102fd16f54b2)

</details>

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Gudahtt committed:
fix: Fix MV2 build sourcemap upload

The sourcemaps for MV2 builds (used for Firefox) were not being
uploaded to Sentry at all. This resulted in invalid stack traces for
Firefox error reports.

The Sentry initiatization has been updated to use a `dist` option,
letting us differentiate between different types of build for the same
version. This is now used to signify which builds are mv2 and which are
mv3. Both distributions are uploaded separately as part of the release
process (for Flask and main builds; we don't have MV2 builds for MMI or
beta).

Fixes #26466

Gudahtt commented on development/sentry-publish.js:
```diff
@@ -109,17 +108,6 @@ async function checkIfVersionExists(version) {
   );
 }
 
-async function checkIfVersionHasArtifacts(version) {
```
From local testing, it seems that this was not working. It was returning empty even for releases with artifacts. I think this is a legacy command that no longer works due to a change in how Sentry tracks artifacts that was made >1 year ago (they're no longer considered as "release bundles", they're "artifact bundles" now, which aren't returned here.)

Since it wasn't working anyway, I removed it rather than trying to update it to use `--dist` correctly (it didn't seem to accept a `--dist` flag, as far as I could tell, so there was no easy way to get this to work).

Gudahtt commented on development/sentry-publish.js:
```diff
@@ -75,19 +79,14 @@ async function start() {
+  }
   if (buildType !== loadBuildTypesConfig().default) {
-    additionalUploadArgs.push('--dist-directory', `dist-${buildType}`);
+    additionalUploadArgs.push('--dist-directory', dist ? `dist-${buildType}-${dist}` : `dist-${buildType}`);
```
This reflects the directory names we use in the CircleCI config

DDDDDanica commented:
LGTM !

Gudahtt committed:
Fix order of script function parameters

Gudahtt commented:
I manually tested this and found one bug, two parameters were in the wrong order. Fixed in be3203f517fb7a85b02bd6e7429743957a0c5a5e

It appears to work correctly now.

davidmurdoch commented on app/scripts/lib/setupSentry.js:
```diff
@@ -477,6 +478,7 @@ function getClientOptions() {
     beforeBreadcrumb: beforeBreadcrumb(),
     beforeSend: (report) => rewriteReport(report),
     debug: METAMASK_DEBUG,
+    dist: isManifestV3 ? ' mv3' : 'mv2',
```
is the space before `mv3` intentional? or the lack of space before `mv2`?

Gudahtt replied:
Good catch, definitely not intentional

```suggestion
    dist: isManifestV3 ? 'mv3' : 'mv2',
```

Gudahtt committed:
Remove accidental leading space in dist

seaona reviewed (dismissed):
code LGTM and the functionality is working. Tested:
- flask mv2
- mv2
- mv3
All source maps are uploaded correctly to Sentry


![Screenshot from 2024-08-20 18-12-26](https://github.com/user-attachments/assets/8d3a66db-d8ba-467c-9ed0-ec00368a2f59)

Gudahtt commented:
Found a problem with the MV3 builds. They're still showing as having a `dist` of `mv3` because `isManifestV3` is coming out as `false` when loaded as part of `setupSentry.js` even on MV3 builds.

Gudahtt committed:
Why is this cursed

Gudahtt committed:
Revert "Why is this cursed"

This reverts commit 4bd50d4c78e2a912d6a23744905da5bf79431894.

Gudahtt committed:
Fix `isManifestV3` value during Sentry setup

I was seeing a bug where MV3 builds were still set as `dist: 'mv2'`.
This was because `isManifestV3` was resolving to `false` in the
`sentry-install.js` bundle for some reason.

It turns out that the runtime manifest is unavailable because the
Chrome extension API doesn't exist in the global scope early on in a
service worker process. It magically appears some time later. So in
this case, the fallback `process.env.ENABLE_MV3` was used. This
fallback case was broken because it expected `ENABLE_MV3` to be set
to the string `"true"`, but it was set to the boolean `true`. The
fallback check has been updated to account for both cases.

Gudahtt commented:
The issue referenced [here](https://github.com/MetaMask/metamask-extension/pull/26467#issuecomment-2299566899) was resolved in [this commit](https://github.com/MetaMask/metamask-extension/pull/26467/commits/011ff476040784a3058f98355ad47a2b8ea97554). This is ready for review again, I have re-tested it. Updating the testing instructions now.

desi reviewed (approved):
I don't see anything here that is alarming from a circle ci point of view and trust that the other reviewers would have caught issues in the other areas I am less familiar with. LGTM.
